File: frontend/src/lib/life-expectancy.ts
import data from "@/data/life_expectancy_2023.json";

export type GenderInput = "m" | "f" | "all" | "male" | "female" | "total";
export type NormalizedGender = "male" | "female" | "total";

export type LifeExpectancyResponse = {
  gender: NormalizedGender;
  referenceYear: number;
  ageInReferenceYear: number;
  remainingYears: number;
  remainingYearsRounded: number;
};

type LifeExpectancyEntry = {
  age: number;
  total: number;
  male: number;
  female: number;
};

type BirthdateParts = {
  year: number;
  month: number;
  day: number;
};

const entriesByAge = new Map<number, LifeExpectancyEntry>(
  data.entries.map((entry) => [entry.age, entry]),
);

const ages = Array.from(entriesByAge.keys());
const MIN_AGE = Math.min(...ages);
const MAX_AGE = Math.max(...ages);

const GENDER_MAP: Record<GenderInput, NormalizedGender> = {
  m: "male",
  male: "male",
  f: "female",
  female: "female",
  all: "total",
  total: "total",
};

function normalizeGender(gender: string): NormalizedGender | null {
  const normalized = GENDER_MAP[gender.toLowerCase() as GenderInput];
  return normalized ?? null;
}

function makeUTCDate(parts: BirthdateParts | { year: number; month: number; day: number }): Date {
  return new Date(Date.UTC(parts.year, parts.month - 1, parts.day));
}

function parseBirthdate(birthdate: string): BirthdateParts | null {
  const match = /^(\d{4})-(\d{2})-(\d{2})$/.exec(birthdate.trim());
  if (!match) return null;

  const [year, month, day] = match.slice(1).map(Number);
  const parsed = makeUTCDate({ year, month, day });
  if (
    parsed.getUTCFullYear() !== year ||
    parsed.getUTCMonth() + 1 !== month ||
    parsed.getUTCDate() !== day
  ) {
    return null;
  }
  return { year, month, day };
}

function getExpectancy(age: number, gender: NormalizedGender): number {
  const targetAge = Math.min(Math.max(age, MIN_AGE), MAX_AGE);
  const entry = entriesByAge.get(targetAge);
  if (!entry) {
    throw new Error(`No life expectancy data available for age ${targetAge}.`);
  }

  if (gender === "male") return entry.male;
  if (gender === "female") return entry.female;
  return entry.total;
}

function ageAsOfReferenceYear(birthdate: BirthdateParts, referenceYear: number, today: Date): number {
  const refMonth = today.getUTCMonth() + 1;
  const refDay = today.getUTCDate();
  const years = referenceYear - birthdate.year - (birthdate.month > refMonth || (birthdate.month === refMonth && birthdate.day > refDay) ? 1 : 0);
  return Math.max(years, 0);
}

function ensureNotFutureDate(birthdate: BirthdateParts, today: Date): void {
  const todayParts = {
    year: today.getUTCFullYear(),
    month: today.getUTCMonth() + 1,
    day: today.getUTCDate(),
  };

  const birthDateObj = makeUTCDate(birthdate);
  const todayDateObj = makeUTCDate(todayParts);

  if (birthDateObj > todayDateObj) {
    throw new Error("Birthdate cannot be in the future.");
  }
}

export function computeLifeExpectancy(
  birthdate: string,
  genderInput: string,
  referenceDate: Date = new Date(),
): LifeExpectancyResponse {
  const parsedBirthdate = parseBirthdate(birthdate);
  if (!parsedBirthdate) {
    throw new Error("Birthdate must be in YYYY-MM-DD format.");
  }

  ensureNotFutureDate(parsedBirthdate, referenceDate);

  const gender = normalizeGender(genderInput);
  if (!gender) {
    throw new Error("Gender must be one of 'm', 'f', or 'all'.");
  }

  const ageInReferenceYear = ageAsOfReferenceYear(parsedBirthdate, data.referenceYear, referenceDate);
  const remainingYears = getExpectancy(ageInReferenceYear, gender);
  const remainingYearsRounded = Math.ceil(remainingYears);

  return {
    gender,
    referenceYear: data.referenceYear,
    ageInReferenceYear,
    remainingYears,
    remainingYearsRounded,
  };
}

export function getAgeBounds() {
  return { minAge: MIN_AGE, maxAge: MAX_AGE };
}

File: life_expectancy.py
from __future__ import annotations

import math
from dataclasses import dataclass
from datetime import date
from typing import Dict


@dataclass(frozen=True)
class LifeExpectancyEntry:
    age: int
    total: float
    male: float
    female: float


@dataclass(frozen=True)
class LifeExpectancyResult:
    gender: str
    reference_year: int
    age_in_reference_year: int
    remaining_years: float
    remaining_years_ceiling: int


class LifeExpectancyTable:
    def __init__(self, entries: Dict[int, LifeExpectancyEntry], reference_year: int = 2023):
        if not entries:
            raise ValueError("At least one entry is required to build the life expectancy table.")
        self._entries = dict(entries)
        self.reference_year = reference_year
        self._min_age = min(self._entries)
        self._max_age = max(self._entries)

    def age_as_of(self, birthdate: date, reference_date: date | None = None) -> int:
        reference_date = reference_date or date.today()
        ref_date = date(self.reference_year, reference_date.month, reference_date.day)
        years = ref_date.year - birthdate.year
        if (birthdate.month, birthdate.day) > (ref_date.month, ref_date.day):
            years -= 1
        return max(years, 0)

    def get_expectancy(self, age: int, gender: str) -> float:
        normalized_gender = gender.lower()
        if normalized_gender not in {"male", "female", "total"}:
            raise ValueError("Gender must be one of 'male', 'female', or 'total'.")

        target_age = min(max(age, self._min_age), self._max_age)
        entry = self._entries.get(target_age)
        if entry is None:
            raise KeyError(f"No life expectancy data available for age {target_age}.")

        if normalized_gender == "male":
            return entry.male
        if normalized_gender == "female":
            return entry.female
        return entry.total


def compute_residual_life_expectancy(
    table: LifeExpectancyTable,
    birthdate: date,
    gender: str,
    reference_date: date | None = None,
) -> LifeExpectancyResult:
    age_in_reference_year = table.age_as_of(birthdate, reference_date)
    remaining_years = table.get_expectancy(age_in_reference_year, gender)
    rounded_remaining = math.ceil(remaining_years)

    return LifeExpectancyResult(
        gender=gender.lower(),
        reference_year=table.reference_year,
        age_in_reference_year=age_in_reference_year,
        remaining_years=remaining_years,
        remaining_years_ceiling=rounded_remaining,
    )
